camera_system/optimal_motion_searcher.py

```diff
@@ -0,0 +1,233 @@
+                      "logs": [copy.deepcopy(start_robot)]}                     # 走行体の推移
+        }
+
+        # 探索対象がブロック設置動作か、ブロック取得動作か(True:設置, False:取得)
+        is_set_motion = goal_node.node_type != NodeType.BLOCK.value
+        # 設置動作探索時、ゴールノードにブロックがある場合
+        if is_set_motion and goal_node.block_id != -1:
+            print("A block already exists at the goal node.")
+            # 空のCompositeGameMotionを返す
+            return CompositeGameMotion()
+
+        # 最適動作を探索する
+        while transition_table[open[0]]["robot"].coord != goal_node.coord:
+            # openの先頭(推定コストが最小)を取り出す
+            min_cost_transition = transition_table[open.pop(0)]
+            # 遷移可能な走行体を取得する
+            next_robots = cls.next_robots(
+                min_cost_transition["robot"], goal_node.coord, is_set_motion)
+
+            # 遷移可能な走行体について探索する
+            for robot in next_robots:
+                # 走行体のハッシュ値を求める
+                hash = cls.robot_hash(robot)
+                # 開始時の走行体 -> 1ゲーム動作前の走行体 の動作群
+                motions = copy.deepcopy(min_cost_transition["motions"])
+                # 1ゲーム動作前の走行体 -> 探索対象の走行体 の動作
+                motion = MotionConverterMock.convert(min_cost_transition["robot"], robot)
+                # 開始時の走行体 -> 探索対象の走行体 の動作群
+                motions.append_game_motion(motion)
+
+                # 運搬対象の走行体に関する探索情報を生成する
+                transition = {"robot": robot,
+                              "motions": motions,
+                              "cost": motions.get_cost() + cls.predict_cost(robot, goal_node.coord),
+                              "logs": min_cost_transition["logs"] + [robot]}
+                # 探索情報がテーブルにない場合は探索情報を登録する
+                if hash not in transition_table.keys():
+                    transition_table[hash] = transition
+                # 探索情報がテーブルにあり、より低コストで遷移できる場合は探索情報更新する
+                elif transition["cost"] < transition_table[hash]["cost"]:
+                    transition_table[hash] = transition
+                # それ以外の場合、探索情報を破棄する
+                else:
+                    continue
+                open.append(hash)
+
+            # 遷移できる走行体がない場合
+            if open == []:
+                print("Impossible move (%d,%d,%s) to (%d,%d)." %
+                      (start_robot.coord.x, start_robot.coord.y, start_robot.direct.name,
+                       goal_node.coord.x, goal_node.coord.y))
+                # 空のCompositeGameMotionを返す
+                return CompositeGameMotion()
+
+            # 最小コストのハッシュ値をリストの先頭に持ってくる
+            list(set(open))
+            min_hash = open[0]
+            for op in open:
+                if transition_table[op]["cost"] < transition_table[min_hash]["cost"]:
+                    min_hash = op
+            open.remove(min_hash)
+            open.insert(0, min_hash)
+
+        # 探索した最適動作を返す
+        return transition_table[open[0]]["motions"]
+
+    @classmethod
+    def predict_cost(cls, start_robot: Robot, goal_coord: Coordinate) -> int:
```

Comment:
この関数ってクラスの外から使わないならプライベートにした方がよくない？
```suggestion
    @classmethod
    def __predict_cost(cls, start_robot: Robot, goal_coord: Coordinate) -> int:
```